### PR TITLE
dts: remove legacy macro support

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
The legacy macros were first deprecated in Zephyr v2.3. Now that
Zephyr v2.4 has been released, that makes two releases where these
macros have been deprecated, so it's OK to remove them.

This leaves support for legacy binding syntax in place. Removing that
is left to future work.

We need to update various pieces of documentation related to flash
partitions that never got updated when the new API was introduced.
Consolidate this information in the flash_map.h API reference page,
since that's really where users will run into it. This also gives us
the opportunity to improve this documentation.

Adjust a couple of kconfigfunctions.py and sanitycheck bits to use
non-legacy edtlib APIs.

### CODE REVIEW

**Comment by galak** [comment]
LGTM.

**Comment by mbolivar-nordic** [comment]
> LGTM.

Thanks. I think all that's left is to rework some docs. We need a user guide for fixed-partitions that isn't on the legacy page.

**Review by galak — APPROVED** [review]

**Review by erwango — COMMENTED** [review]
I'm fine globally, but one question though:
Should we still generate device_tree_fixup.h ?

**Comment by mbolivar-nordic** [comment]
> Should we still generate device_tree_fixup.h ?

That's a separate question from the legacy macros, no? We can decide whether to remove fixups and legacy binding support apart from this PR.

For me at least, fixups and legacy bindings have caused zero maintenance burden, so I think it's less work just to leave them in for now, unless they're causing people a problem.

It is nice to get rid of gen_legacy_defines.py, since that has required a bit of maintenance here and there and we've had to be concerned about maintaining compatibility.

**Review by erwango — APPROVED** [review]

File: include/storage/flash_map.h [hunk 1]
```diff
@@ -46,22 +39,32 @@
...
+ * This structure represents a fixed-size partition on a flash device.
+ * Each partition contains one or more flash sectors.
  */
 struct flash_area {
-	uint8_t fa_id; /** ID of flash area */
+	/** ID number */
+	uint8_t fa_id;
+	/** Provided for compatibility with MCUboot */
 	uint8_t fa_device_id;
 	uint16_t pad16;
```
Comment (nvlsianpu) [inline include/storage/flash_map.h:58]: I doubt that `pad16` is required for anything.

File: doc/guides/dts/howtos.rst [hunk 2]
```diff
@@ -566,218 +566,6 @@ done in the :ref:`blinky-sample`. The application can then be configured in
 :ref:`BOARD.dts <devicetree-in-out-files>` files or via :ref:`devicetree
 overlays <use-dt-overlays>`.
 
-.. _dt-migrate-legacy:
```
[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply doc/guides/dts/howtos.rst:569]: > I would kept migration guide with information that legacy macros were removed completely after 2.4 releases.

I did, in the release notes.

File: include/storage/flash_map.h (hunk 1 above)
Comment (mbolivar-nordic) [reply include/storage/flash_map.h:58]: > I doubt that `pad16` is required for anything.

Are you asking for it to be removed? That's an API change not relevant to this PR.

File: doc/reference/storage/flash_map/flash_map.rst [hunk 3]
```diff
@@ -1,36 +1,92 @@
...
-location in flash_area structures array form.
+The ``<storage/flash_map.h>`` API allows accessing information about device
+flash partitions via :c:struct:`flash_area` structures.
 
-Flash map is generated from DTS based on content of :ref:`legacy_flash_partitions`
-nodes.
-The flash_area API provides a way to access data in the flash map.
-The flash_area_open() API is the interface for obtaining the flash partitions
-flash_area from the flash map.
+Each ``struct flash_area`` describes a flash partition and has a globally
```
[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply doc/reference/storage/flash_map/flash_map.rst]: > flash_area API might be used also standalone as abstraction for convenient access to any flash memory region.

Reworded to:

```
Each ``struct flash_area`` describes a flash partition. The API provides access
to a "flash map", which contains predefined flash areas accessible via globally
unique ID numbers. You can also create ``flash_area`` structures at runtime for
application-specific purposes.

The ``flash_area`` structure contains the name of the flash device the
partition is part of; this name can be passed to :c:func:`device_get_binding`
to get the corresponding :c:struct:`device` structure which can be read and
written to using the :ref:`flash API <flash_api>`. The ``flash_area`` also
contains the start offset and size of the partition within the flash memory the
device represents.
```

**Comment by carlescufi** [comment]
@nvlsianpu please take another look

File: include/storage/flash_map.h (hunk 1 above)
Comment (nvlsianpu) [reply include/storage/flash_map.h:58]: sure - this is somethin i can do by myself.

File: doc/guides/dts/howtos.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply doc/guides/dts/howtos.rst:569]: forgot to said that I encounter this record after I had write down the comment.
Yes, that record should be enough.

**Review by nvlsianpu — APPROVED** [review]

**Pull request merged**